src/Arrays/MPIStateArrays.jl

```diff
@@ -756,9 +791,11 @@ end
 # better names, for example `device` is also defined in CUDAdrv
 
 device(::Union{Array, SArray, MArray}) = CPU()
+device(Q::MPIDestArray) = device(parent(Q))
```

Comment:
```
for (W, _) in Adapt.wrappers
    @eval device(wrapper::$W where {AT <: Any}) = device(parent(wrapper))
end
```

Instead of specialising on the Union.